Update Apex Writing and Testing (#1931)
forcedotcom/salesforcedx-vscode: docs/articles/apex/writing.md

title: Write Apex Code
---

Use the Apex extension for VS Code to access code-editing features such as code completion, go to definition, view outlines of Apex classes and triggers, refactoring, and find syntactic errors in your code.

## Code-Completion

> Apex extension provides context-sensitive suggestions when you’re working in an Apex class or trigger. As you type, the auto-completion list of members such as, methods or variables pop up. The auto-completion list also displays the documentation for the suggestions. Use these keys:

Reviewer: Do we need to start the paragraphs in each section of this doc with "Apex extension"? For example, just say "Context-sensitive suggestions can be provided when working in an Apex class or trigger" or for below, "Code snippets are available for scaffolding class and interface definitions, and a variety of statements such as loops, conditionals, etc."

Author: Updated so that all the sections don't repeat 'Apex extension ...'

- Ctrl+space to view see code-completion suggestions
- Arrow keys to navigate between the suggestions
- Enter key to select from the suggestions

![Animation showing code completion of PropertyController](../../images/apex_completion_with_doc.gif)

> If you want to change how suggestions are pre-selected, see [IntelliSense - Suggestion selection](https://code.visualstudio.com/docs/editor/intellisense).

Reviewer: ```suggestion
If you want to change how suggestions are pre-selected, see [IntelliSense - Suggestion selection](https://code.visualstudio.com/docs/editor/intellisense#_customizing-intellisense).
```

Minor change but this will take folks directly to the section on how to customize intellisense.

  - Variables (local and class variables)
- Standard objects
  - Fields (standard and custom fields)
  - Object definition
- Custom objects
  - Fields
  - Object definitions

> When you refresh SObject definitions, the extension uses the default org to generate Apex classes. These classes represent the standard and custom objects that the current user has access to. **Do not** edit these representative Apex classes, which are stored in a hidden directory on your local workstation.

Reviewer: ```suggestion
When you refresh SObject definitions, the extension uses the default org to generate Apex classes. These classes represent the standard and custom objects that the current user has access to. **Do not** edit these representative Apex classes, which are stored in the hidden `.sfdx` directory in your project.
```

Reviewer: I'm torn about saying "__do not__" edit these classes if the user knows what they're doing. I'd maybe say it isn't recommended and explain that they're intended to be read-only and if they're modified, it may impact the autocomplete experience and any changes made may be wiped when rerunning the refresh command.

Whenever you refresh SObject definitions, the representative Apex classes are deleted and regenerated. You can modify the SObjects either by updating the objects’ `.object-meta.xml` and `.field-meta.xml` files; Or by making changes declaratively in the default org. After modifying the SObjects, make sure to sync your local project and the default org.

> When you launch the Salesforce CLI Integration extension (which is part of the Salesforce Extension Pack) for the first time, `SFDX: Refresh SObject Definitions` command is executed automatically if your project doesn’t contain any Apex classes. After you add or edit standard or custom objects or their fields, make sure to run this command.

Reviewer 2: This is currently not true. For sobject refresh to run automatically the user needs to enable the `salesforcedx-vscode-apex.enable-sobject-refresh-on-startup` setting.

- To preview a definition, press and hold Ctrl (Windows or Linux) or Cmd (macOS) and hover over the item whose definition you want to see.
- To view a definition, right-click the item and select **Peek Definition**, or press Alt+F12.
- To jump to the location of a definition, right-click the item and select **Go to Definition**, or press F12.

![Previewing, viewing, and jumping to a definition](../../images/apex_go_to_definition.gif)

## Find All References

You can find all references to user-defined Apex such as Classes, Class variables, Enums, Interfaces, Methods, and Properties. To find references, select an item and press Shift+F12; Or right-click the item and select **Find All References**. The reference results are displayed in the left pane of the editor window.

## Outline View

The Apex outline view shows the structure of the Apex class or trigger that’s open in the editor.

- To view the list of the symbols in the file, press Ctrl+Shift+O (Windows or Linux) or Cmd+Shift+O (macOS).
- To go to one of the symbols, select it from the list.

> ![Outline view, showing the symbols in an Apex class](../../images/apex_outline.png)

Reviewer: We should also point out that there's also an actual outline section in the Explorer View
![Screen Shot 2020-02-06 at 11 09 59 AM](https://user-images.githubusercontent.com/10779873/73970044-546cc280-48d1-11ea-9af8-7cc0ba3efb1b.png)

Author: Updated the section with this info.

Author: @brpowell, @smaddox-sf , @lcampos , @sfsholden, Addressed all the review comments in Apex Writing and Testing topics. Please do let me know if you have any further feedback for these 2 topics.
Also, I created this branch based on an outdated branch and not the develop branch (facepalm). So I'll delete this pull request and create a new one...
